Build failures and internal errors when switching from 7.0.0rc3 to 7.0.0rc4

### Description of the bug:


Bazel 7.0.0rc3 was building without problems our big monorepo, but 7.0.0rc4 fails with errors, e.g.:
```
WARNING: /Users/rsalvador/.../BUILD.bazel:3:17: output 'ui-services-connection-models-api/modelbuilder_generated' of //ui-services-connection-models-api:ui_services_connection_model is a directory; dependency checking of directories is unsound
ERROR: /Users/rsalvador/.../BUILD.bazel:3:17: modelbuilder generating srcjar modelbuilder_generated.srcjar failed: Exec failed due to IOException: /Users/rsalvador/.cache/bazel/b6de016d545fc335c6fe85486d422c54/execroot/core/bazel-out/darwin_x86_64-fastbuild/bin/ui-services-connection-models-api/modelbuilder_generated (No such file or directory)
```

and

```
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'ActionLookupData6{actionLookupKey=ConfiguredTargetKey{label=//tools/build/bazel/rules/lwc/js:bundle, config=BuildConfigurationKey[6aa701ced8cd3692bfda0c63eeb4de6bb53c8649464bf1688db8c6f5976ed915]}, actionIndex=6}' (requested by nodes 'ArtifactNestedSetKey[6]@774500339', 'ArtifactNestedSetKey[5]@1662963755')
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:550)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.UnsupportedOperationException: Cannot get node id for RegularFileArtifactValue{digest=e58e4f6d4f57ab1e4a503a523d1c4e34ed03eee292a0237a898ac158c1b7bc69, size=1080, proxy=ctime of 1700240624111 and mtime of 499162500000 and nodeId of 2582129255}
	at com.google.devtools.build.lib.remote.RemoteActionFileSystem$2.getNodeId(RemoteActionFileSystem.java:634)
	at com.google.devtools.build.lib.vfs.DigestUtils$CacheKey.<init>(DigestUtils.java:67)
	at com.google.devtools.build.lib.vfs.DigestUtils.manuallyComputeDigest(DigestUtils.java:193)
	at com.google.devtools.build.lib.vfs.DigestUtils.getDigestWithManualFallback(DigestUtils.java:160)
	at com.google.devtools.build.lib.exec.SpawnLogContext.computeDigest(SpawnLogContext.java:348)
	at com.google.devtools.build.lib.exec.SpawnLogContext.listDirectoryContents(SpawnLogContext.java:296)
	at com.google.devtools.build.lib.exec.SpawnLogContext.logSpawn(SpawnLogContext.java:130)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:192)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:119)
	at com.google.devtools.build.lib.exec.SpawnStrategyResolver.exec(SpawnStrategyResolver.java:45)
	at com.google.devtools.build.lib.analysis.actions.SpawnAction.execute(SpawnAction.java:261)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.executeAction(SkyframeActionExecutor.java:1148)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1065)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:165)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:94)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:562)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:859)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:333)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:171)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
```

the first error we can probably fix in the generator, but the internal error may point to some regression?

The internal error goes away if we don't use the `--execution_log_json_file` and `--noexecution_log_sort` flags.

### Which category does this issue belong to?


Core

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.


We can't provide a minimal example, hit happens deep into the build of a very big monorepo.

### Which operating system are you running Bazel on?


MacOS Version 14.1

### What is the output of `bazel info release`?


release 7.0.0rc4

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.


_No response_

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?


_No response_

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.


Yes, it is a regression.

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

--- a/src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java
+++ b/src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java
@@ -349,16 +349,22 @@ public class RemoteActionFileSystem extends AbstractFileSystemWithCustomStat {
 
   @Override
   protected byte[] getFastDigest(PathFragment path) throws IOException {
-    var stat = statInMemory(path, FollowMode.FOLLOW_ALL);
-    if (stat instanceof FileStatusWithDigest) {
-      return ((FileStatusWithDigest) stat).getDigest();
+    path = resolveSymbolicLinks(path).asFragment();
+    // The parent path has already been canonicalized by resolveSymbolicLinks, so FOLLOW_NONE is
+    // effectively the same as FOLLOW_PARENT, but more efficient.
+    var status = statInMemory(path, FollowMode.FOLLOW_NONE);
+    if (status instanceof FileStatusWithDigest) {
+      return ((FileStatusWithDigest) status).getDigest();
     }
     return localFs.getPath(path).getFastDigest();
   }
 
   @Override
   protected byte[] getDigest(PathFragment path) throws IOException {
-    var status = statInMemory(path, FollowMode.FOLLOW_ALL);
+    path = resolveSymbolicLinks(path).asFragment();
+    // The parent path has already been canonicalized by resolveSymbolicLinks, so FOLLOW_NONE is
+    // effectively the same as FOLLOW_PARENT, but more efficient.
+    var status = statInMemory(path, FollowMode.FOLLOW_NONE);
     if (status instanceof FileStatusWithDigest) {
       return ((FileStatusWithDigest) status).getDigest();
     }
@@ -566,6 +572,7 @@ public class RemoteActionFileSystem extends AbstractFileSystemWithCustomStat {
     if (status != null) {
       return status;
     }
+
     // The path has already been canonicalized above.
     return localFs.getPath(path).stat(Symlinks.NOFOLLOW);
   }

--- a/src/main/java/com/google/devtools/build/lib/testing/vfs/SpiedFileSystem.java
+++ b/src/main/java/com/google/devtools/build/lib/testing/vfs/SpiedFileSystem.java
@@ -58,6 +58,11 @@ public class SpiedFileSystem extends DelegateFileSystem {
   }
 
   @Override
+  public byte[] getDigest(PathFragment path) throws IOException {
+    return super.getDigest(path);
+  }
+
+  @Override
   public void chmod(PathFragment path, int mode) throws IOException {
     super.chmod(path, mode);
   }

--- a/src/test/java/com/google/devtools/build/lib/remote/BUILD
+++ b/src/test/java/com/google/devtools/build/lib/remote/BUILD
@@ -108,6 +108,7 @@ java_library(
         "//src/main/java/com/google/devtools/build/lib/remote/util",
         "//src/main/java/com/google/devtools/build/lib/runtime/commands",
         "//src/main/java/com/google/devtools/build/lib/skyframe:tree_artifact_value",
+        "//src/main/java/com/google/devtools/build/lib/testing/vfs:spied_filesystem",
         "//src/main/java/com/google/devtools/build/lib/util:abrupt_exit_exception",
         "//src/main/java/com/google/devtools/build/lib/util:exit_code",
         "//src/main/java/com/google/devtools/build/lib/util:os",

--- a/src/test/java/com/google/devtools/build/lib/remote/RemoteActionFileSystemTest.java
+++ b/src/test/java/com/google/devtools/build/lib/remote/RemoteActionFileSystemTest.java
@@ -25,12 +25,14 @@ import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.doReturn;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.reset;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.verifyNoMoreInteractions;
 
+import com.google.common.base.Utf8;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
-import com.google.common.hash.HashCode;
 import com.google.common.util.concurrent.ListenableFuture;
 import com.google.devtools.build.lib.actions.ActionExecutionMetadata;
 import com.google.devtools.build.lib.actions.ActionInput;
@@ -50,7 +52,7 @@ import com.google.devtools.build.lib.actions.util.ActionsTestUtil;
 import com.google.devtools.build.lib.clock.JavaClock;
 import com.google.devtools.build.lib.remote.options.RemoteOutputsMode;
 import com.google.devtools.build.lib.skyframe.TreeArtifactValue;
-import com.google.devtools.build.lib.vfs.DigestHashFunction;
+import com.google.devtools.build.lib.testing.vfs.SpiedFileSystem;
 import com.google.devtools.build.lib.vfs.Dirent;
 import com.google.devtools.build.lib.vfs.FileStatus;
 import com.google.devtools.build.lib.vfs.FileStatusWithDigest;
@@ -62,7 +64,6 @@ import com.google.devtools.build.lib.vfs.PathFragment;
 import com.google.devtools.build.lib.vfs.Root;
 import com.google.devtools.build.lib.vfs.Symlinks;
 import com.google.devtools.build.lib.vfs.SyscallCache;
-import com.google.devtools.build.lib.vfs.inmemoryfs.InMemoryFileSystem;
 import com.google.errorprone.annotations.CanIgnoreReturnValue;
 import com.google.testing.junit.testparameterinjector.TestParameter;
 import com.google.testing.junit.testparameterinjector.TestParameterInjector;
@@ -81,12 +82,10 @@ public final class RemoteActionFileSystemTest extends RemoteActionFileSystemTest
       new RemoteOutputChecker(
           new JavaClock(), "build", RemoteOutputsMode.MINIMAL, ImmutableList.of());
 
-  private static final DigestHashFunction HASH_FUNCTION = DigestHashFunction.SHA256;
-
   private static final String RELATIVE_OUTPUT_PATH = "out";
 
   private final RemoteActionInputFetcher inputFetcher = mock(RemoteActionInputFetcher.class);
-  private final FileSystem fs = new InMemoryFileSystem(HASH_FUNCTION);
+  private final SpiedFileSystem fs = SpiedFileSystem.createInMemorySpy();
   private final Path execRoot = fs.getPath("/exec");
   private final ArtifactRoot sourceRoot = ArtifactRoot.asSourceRoot(Root.fromPath(execRoot));
   private final ArtifactRoot outputRoot =
@@ -459,6 +458,91 @@ public final class RemoteActionFileSystemTest extends RemoteActionFileSystemTest
 
     assertThrows(
         FileNotFoundException.class, () -> actionFs.stat(path, /* followSymlinks= */ true));
+  }
+
+  @Test
+  public void getDigest_fromInputArtifactData_forLocalArtifact() throws Exception {
+    ActionInputMap inputs = new ActionInputMap(1);
+    Artifact artifact = createRemoteArtifact("file", "local contents", inputs);
+    PathFragment path = artifact.getPath().asFragment();
+    RemoteActionFileSystem actionFs = (RemoteActionFileSystem) createActionFileSystem(inputs);
+
+    // Verify that we don't fall back to a slow digest.
+    reset(fs);
+    assertThat(actionFs.getFastDigest(path)).isEqualTo(getDigest("local contents"));
+    verify(fs, never()).getDigest(any());
+
+    assertThat(actionFs.getDigest(path)).isEqualTo(getDigest("local contents"));
+  }
+
+  @Test
+  public void getDigest_fromInputArtifactData_forRemoteArtifact() throws Exception {
+    ActionInputMap inputs = new ActionInputMap(1);
+    Artifact artifact = createRemoteArtifact("file", "remote contents", inputs);
+    PathFragment path = artifact.getPath().asFragment();
+    RemoteActionFileSystem actionFs = (RemoteActionFileSystem) createActionFileSystem(inputs);
+
+    // Verify that we don't fall back to a slow digest.
+    reset(fs);
+    assertThat(actionFs.getFastDigest(path)).isEqualTo(getDigest("remote contents"));
+    verify(fs, never()).getDigest(any());
+
+    assertThat(actionFs.getDigest(path)).isEqualTo(getDigest("remote contents"));
+  }
+
+  @Test
+  public void getDigest_fromRemoteOutputTree() throws Exception {
+    RemoteActionFileSystem actionFs = (RemoteActionFileSystem) createActionFileSystem();
+    Artifact artifact = ActionsTestUtil.createArtifact(outputRoot, "out");
+    PathFragment path = artifact.getPath().asFragment();
+    injectRemoteFile(actionFs, artifact.getPath().asFragment(), "remote contents");
+
+    assertThat(actionFs.getFastDigest(path)).isEqualTo(getDigest("remote contents"));
+    assertThat(actionFs.getDigest(path)).isEqualTo(getDigest("remote contents"));
+  }
+
+  @Test
+  public void getDigest_fromLocalFilesystem() throws Exception {
+    RemoteActionFileSystem actionFs = (RemoteActionFileSystem) createActionFileSystem();
+    Artifact artifact = ActionsTestUtil.createArtifact(outputRoot, "out");
+    PathFragment path = artifact.getPath().asFragment();
+    writeLocalFile(actionFs, artifact.getPath().asFragment(), "local contents");
+
+    assertThat(actionFs.getFastDigest(path)).isNull();
+    assertThat(actionFs.getDigest(path)).isEqualTo(getDigest("local contents"));
+  }
+
+  @Test
+  public void getDigest_notFound() throws Exception {
+    RemoteActionFileSystem actionFs = (RemoteActionFileSystem) createActionFileSystem();
+    Artifact artifact = ActionsTestUtil.createArtifact(outputRoot, "out");
+    PathFragment path = artifact.getPath().asFragment();
+
+    assertThrows(FileNotFoundException.class, () -> actionFs.getFastDigest(path));
+    assertThrows(FileNotFoundException.class, () -> actionFs.getDigest(path));
+  }
+
+  @Test
+  public void getDigest_followSymlinks(
+      @TestParameter FilesystemTestParam from, @TestParameter FilesystemTestParam to)
+      throws Exception {
+    RemoteActionFileSystem actionFs = (RemoteActionFileSystem) createActionFileSystem();
+    FileSystem fromFs = from.getFilesystem(actionFs);
+    FileSystem toFs = to.getFilesystem(actionFs);
+
+    PathFragment linkPath = getOutputPath("sym");
+    PathFragment targetPath = getOutputPath("target");
+    fromFs.getPath(linkPath).createSymbolicLink(execRoot.getRelative(targetPath).asFragment());
+
+    if (toFs.equals(actionFs.getLocalFileSystem())) {
+      writeLocalFile(actionFs, targetPath, "content");
+      assertThat(actionFs.getFastDigest(linkPath)).isNull();
+    } else {
+      injectRemoteFile(actionFs, targetPath, "content");
+      assertThat(actionFs.getFastDigest(linkPath)).isEqualTo(getDigest("content"));
+    }
+
+    assertThat(actionFs.getDigest(linkPath)).isEqualTo(getDigest("content"));
   }
 
   @Test
@@ -943,16 +1027,12 @@ public final class RemoteActionFileSystemTest extends RemoteActionFileSystemTest
   @CanIgnoreReturnValue
   protected FileArtifactValue injectRemoteFile(
       FileSystem actionFs, PathFragment path, String content) throws IOException {
-    byte[] contentBytes = content.getBytes(UTF_8);
-    HashCode hashCode = HASH_FUNCTION.getHashFunction().hashBytes(contentBytes);
+    byte[] digest = getDigest(content);
+    int size = Utf8.encodedLength(content);
     ((RemoteActionFileSystem) actionFs)
-        .injectRemoteFile(
-            path, hashCode.asBytes(), contentBytes.length, /* expireAtEpochMilli= */ -1);
+        .injectRemoteFile(path, digest, size, /* expireAtEpochMilli= */ -1);
     return RemoteFileArtifactValue.create(
-        hashCode.asBytes(),
-        contentBytes.length,
-        /* locationIndex= */ 1,
-        /* expireAtEpochMilli= */ -1);
+        digest, size, /* locationIndex= */ 1, /* expireAtEpochMilli= */ -1);
   }
 
   @Override
@@ -964,13 +1044,14 @@ public final class RemoteActionFileSystemTest extends RemoteActionFileSystemTest
 
   /** Returns a remote artifact and puts its metadata into the action input map. */
   private Artifact createRemoteArtifact(
-      String pathFragment, String contents, ActionInputMap inputs) {
+      String pathFragment, String content, ActionInputMap inputs) {
     Artifact a = ActionsTestUtil.createArtifact(outputRoot, pathFragment);
-    byte[] b = contents.getBytes(UTF_8);
-    HashCode h = HASH_FUNCTION.getHashFunction().hashBytes(b);
     RemoteFileArtifactValue f =
         RemoteFileArtifactValue.create(
-            h.asBytes(), b.length, /* locationIndex= */ 1, /* expireAtEpochMilli= */ -1);
+            getDigest(content),
+            Utf8.encodedLength(content),
+            /* locationIndex= */ 1,
+            /* expireAtEpochMilli= */ -1);
     inputs.putWithNoDepOwner(a, f);
     return a;
   }
@@ -989,11 +1070,13 @@ public final class RemoteActionFileSystemTest extends RemoteActionFileSystemTest
     TreeArtifactValue.Builder builder = TreeArtifactValue.newBuilder(a);
     for (Map.Entry<String, String> entry : contentMap.entrySet()) {
       TreeFileArtifact child = TreeFileArtifact.createTreeOutput(a, entry.getKey());
-      byte[] b = entry.getValue().getBytes(UTF_8);
-      HashCode h = HASH_FUNCTION.getHashFunction().hashBytes(b);
+      String content = entry.getValue();
       RemoteFileArtifactValue childMeta =
           RemoteFileArtifactValue.create(
-              h.asBytes(), b.length, /* locationIndex= */ 0, /* expireAtEpochMilli= */ -1);
+              getDigest(content),
+              Utf8.encodedLength(content),
+              /* locationIndex= */ 0,
+              /* expireAtEpochMilli= */ -1);
       builder.putChild(child, childMeta);
     }
     return builder.build();
@@ -1044,5 +1127,9 @@ public final class RemoteActionFileSystemTest extends RemoteActionFileSystemTest
       builder.putChild(child, childMeta);
     }
     return builder.build();
+  }
+
+  private byte[] getDigest(String content) {
+    return fs.getDigestFunction().getHashFunction().hashString(content, UTF_8).asBytes();
   }
 }

